pebblo_saferetriever/langchain/identity-rag/pebblo_semantic_rag.py

```diff
@@ -0,0 +1,179 @@
+                file_loader_kwargs={"mode": "elements"},
+                load_auth=True,
+            ),
+            name=self.app_name,  # App name (Mandatory)
+            owner="Joe Smith",  # Owner (Optional)
+            description="Identity enabled SafeLoader and SafeRetrival app using Pebblo",  # Description (Optional)
+            api_key=os.environ.get("PEBBLO_API_KEY"),  # Pebblo cloud API key
+            load_semantic=True,
+        )
+        self.documents = self.loader.load()
+        unique_identities = set()
+        unique_topics = set()
+        unique_entities = set()
+        for doc in self.documents:
+            if doc.metadata.get("authorized_identities"):
+                unique_identities.update(doc.metadata.get("authorized_identities"))
+            if doc.metadata.get("pebblo_semantic_topics"):
+                unique_topics.update(doc.metadata.get("pebblo_semantic_topics"))
+            if doc.metadata.get("pebblo_semantic_entities"):
+                unique_entities.update(doc.metadata.get("pebblo_semantic_entities"))
+
+        print(f"Authorized Identities: {list(unique_identities)}")
+        print(f"Semantic Topics: {list(unique_topics)}")
+        print(f"Semantic Entities: {list(unique_entities)}")
+        print(f"Loaded {len(self.documents)} documents ...\n")
+
+        # Load documents into VectorDB
+
+        print("Hydrating Vector DB ...")
+        self.vectordb = self.embeddings()
+        print("Finished hydrating Vector DB ...\n")
+
+        # Prepare LLM
+        self.llm = OpenAI()
+
+    def embeddings(self):
+        embeddings = OpenAIEmbeddings()
+        vectordb = Qdrant.from_documents(
+            self.documents,
+            embeddings,
+            location=":memory:",
+            collection_name=self.collection_name,
+        )
+        return vectordb
+
+    def ask(
+        self,
+        question: str,
+        auth_identifiers: Optional[List[str]] = None,
+        topics_to_deny: Optional[List[str]] = None,
+        entities_to_deny: Optional[List[str]] = None,
+    ):
+        auth_context = dict()
+        if auth_identifiers:
+            auth_context["authorized_identities"] = auth_identifiers
+        semantic_context = dict()
+        if topics_to_deny:
+            semantic_context["pebblo_semantic_topics"] = {"deny": topics_to_deny}
+        if entities_to_deny:
+            semantic_context["pebblo_semantic_entities"] = {"deny": entities_to_deny}
+
+        auth_context = AuthContext(**auth_context) if auth_context else None
+        semantic_context = (
+            SemanticContext(**semantic_context) if semantic_context else None
+        )
+        # Prepare retriever QA chain
+        retriever = PebbloRetrievalQA.from_chain_type(
+            llm=self.llm,
+            chain_type="stuff",
+            retriever=self.vectordb.as_retriever(),
+            verbose=True,
+            auth_context=auth_context,
+            semantic_context=semantic_context,
```

Comment:
Is it possible to hide complexity of preparing semantic_context? It will reduce number of lines RAG app developer has to write to include semantic enforcement. What if we pass `topics_to_deny` and `entities_to_deny` to ` PebbloRetrievalQA.from_chain_type()`?